File: Mat/branch_update.m
function B_new = branch_update(B,n,l_task,s_task,d_task)
%%% Generate New Branch

seq_prev = B.seq;
t_ex_prev = B.t_ex;
l_inc_prev = B.l_inc;


seq = [seq_prev; n];

t_ex = t_ex_prev;
if isempty(seq_prev)
    t_ex(n) = s_task(n);
else
    t_ex(n) = max([s_task(n) ; t_ex_prev(seq_prev(end)) + d_task(seq_prev(end))]);
end


l_inc = l_inc_prev + l_task{n}(t_ex(n));


LB = l_inc;
UB = l_inc;

N = numel(l_task);
T_c = setdiff((1:N)',seq);

t_end = t_ex(n) + d_task(n);
t_max = sum(d_task(T_c)) + max([s_task(T_c) ; t_end]);

for n = T_c'
    LB = LB + l_task{n}(max([s_task(n) ; t_end]));
    UB = UB + l_task{n}(t_max);
end


B_new = struct('seq',seq,'t_ex',t_ex,'l_inc',l_inc,'LB',LB,'UB',UB);

File: Py/branch_update.py
"""
Function for generating new branches on a Branch-and-Bound tree search.
"""


def branch_update(branch, n, tasks):
    """Generate New Branch"""

    seq_prev = branch['seq']
    t_ex_prev = branch['t_ex']
    l_inc_prev = branch['l_inc']

    seq = seq_prev + [n]

    t_ex = t_ex_prev.copy()
    if len(seq_prev) == 0:
        t_ex[n] = tasks[n].t_start
    else:
        t_ex[n] = max(tasks[n].t_start, t_ex_prev[seq_prev[-1]] + tasks[seq_prev[-1]].duration)

    l_inc = l_inc_prev + tasks[n].loss_fcn(t_ex[n])

    N = len(tasks)
    T_c = list(set(range(N)) - set(seq))

    t_end = t_ex[n] + tasks[n].duration
    t_s_max = max([tasks[i].t_start for i in T_c] + [t_end]) + sum([tasks[i].duration for i in T_c])

    lb = l_inc
    ub = l_inc
    for n in T_c:
        lb += tasks[n].loss_fcn(max(tasks[n].t_start, t_end))
        ub += tasks[n].loss_fcn(t_s_max)

    return {'seq': seq, 't_ex': t_ex, 'l_inc': l_inc, 'LB': lb, 'UB': ub}